test/models/User.spec.ts

```diff
@@ -180,4 +178,94 @@ describe("User", () => {
       expect(profile.connections[0].name).toBe("google-oauth2");
     });
   });
+
+  describe("validate authenctication code", () => {
+    it("should throw a NoCodeError if a user tries to validate a code but no code is stored", async () => {
+      const caller = createCaller({
+        get: async (key: string) => {
+          switch (key) {
+            case "authentication-code":
+              return null;
+          }
+        },
+      });
+
+      try {
+        await caller.validateAuthenticationCode({ code: "123456", email: 'test@example.com', tenantId: 'tenantId' });
+        throw new Error('Should throw')
+      } catch (err: any) {
+        if (err.message !== 'No code found') {
+          throw new Error('Should throw NoCodeError')
```

Comment:
hmmm, I don't like this way of testing exceptions :smile: 

I have flashbacks of the nightmare of testing exceptions in Jest.  It's easy to get false positives everywhere

Reply:
I remember doing things like this https://stackoverflow.com/questions/46042613/how-to-test-the-type-of-a-thrown-exception-in-jest

```
test("Test description", () => {
  const t = () => {
    throw new TypeError();
  };
  expect(t).toThrow(TypeError);
});
```

You have to pass a function into the expect.

Reply:
I always had issues with this in mocha so just did the brute for solution :)

Reply:
So, with jest and async methods this seems to work:
`      expect(caller.validateAuthenticationCode({ code: "123456", email: 'test@example.com', tenantId: 'tenantId' })).rejects.toThrow(NoCodeError)
`
It's a bit neater for sure.

Reply:
Ouch.. Doesn't really work as it bubbles up through tRPC:
```
Expected constructor: NoCodeError
Received constructor: TRPCError
```
Think tRPC repackages this in the correct error in the client. But we don't get that in the test.

Reply:
Bah... this might be prettier but it's worse in every other way. It's a lot harder to debug as you can't put any breakpoints. Not sure there's a benefit to changing it?

Reply:
Follow your heart Markus :smile: 

The issue with the test code you had before is if your code *doesn't* throw an exception, then the tests will pass.  That's the issue with having assertions inside a catch block. To Jest, no news is good news :wink: 

You're not explicitly telling Jest for an exception to be thrown, you're just saying _well if there is an exception, it should be this_

I've seen this cause errors before, but it's not the end of the world

The syntax totally sucks though and I've always had issues with the exceptions bubbling back up :joy:

Reply:
Got it working now.. The original code had a throw after the function call so it would fail if it didn't throw, but.. this is more elegant.